### PR TITLE
Fetch

### PR DESCRIPTION
**Pull request by jasonterando**
*Issue #, if available:* 585

*Description of changes:*  Add fetch patch, including support for built-in fetch API for NodeJS current versions and node-fetch for older NodeJS versions.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/package.json
Comment (carolabadeer) [inline packages/core/package.json:3]: This corresponds to the [current release version (3.5.0)](https://github.com/aws/aws-xray-sdk-node/releases/tag/aws-xray-sdk-node%403.5.0) and should never be changed to a prior version - this version number is only modified by the X-Ray team when preparing a new release. I imagine this is what is causing most of the CI failures. 

To fix this, please keep this version as is - 3.5.0. We have an [sdk_contrib directory](https://github.com/aws/aws-xray-sdk-node/tree/master/sdk_contrib) that houses all community contributed instrumentations, so it would be great if this instrumentation could be moved there. 

Also thought it would be worth mentioning since we do not usually maintain instrumentations in sdk_contrib, would you be willing to maintain it going forward and keep track of any issues that come in?

File: package.json
Comment (carolabadeer) [inline package.json:24]: Following up to the comment on moving to sdk_contrib, the dependencies that are unique to this instrumentation would then be placed in the package.json of the new instrumentation package in sdk_contrib. 
You can reference this [fastify instrumentation PR](https://github.com/aws/aws-xray-sdk-node/pull/523) as a guide to help

File: packages/core/lib/patchers/fetch_p.js
Comment (carolabadeer) [inline packages/core/lib/patchers/fetch_p.js:9]: nit: use const instead of var

File: packages/core/package.json
Comment (carolabadeer) [inline packages/core/package.json:35]: mocha, tsd, and typescript are not needed here since they get pulled from the [root package.json](https://github.com/aws/aws-xray-sdk-node/blob/master/package.json#L44C24-L52)

**Review by carolabadeer — CHANGES_REQUESTED** [review]

**Comment by jasonterando** [comment]
Hi, my wife and I are on vacation for our 25th anniversary. I had to leave the computer at home lol. I will be able to knock these out at the end of the month, sorry for the delay.

File: packages/core/package.json
Comment (jasonterando) [reply packages/core/package.json:3]: Q:  Do you want to create a `patchers` directory in sdk_contrib for this, since all of the existing contributions are all middleware hooks?

**Comment by jasonterando** [comment]
I'm going to try and resurrect this.  Let me know if this seems feasible or if I should abandon.  Thanks!

**Comment by jasonterando** [comment]
> Hi, can you address @carolabadeer's comment: [#590 (comment)](https://github.com/aws/aws-xray-sdk-node/pull/590#discussion_r1194387982)

Done and committed.

File: sdk_contrib/fetch/lib/fetch_p.js
Comment (jj22ee) [inline sdk_contrib/fetch/lib/fetch_p.js:23]: This function seems unneeded, may introduce confusion. Without looking in this captureFetch function, a user may not understand if they are capturing the Fetch module or the Fetch in the global instance. A user would probably have an idea which one they want to capture explicity, so I think just keeping `captureFetchModule` and `captureFetchGlobal` would be fine. Otherwise, the version of Fetch that is actually instrumented is a bit abstracted.

Comment (jasonterando) [reply sdk_contrib/fetch/lib/fetch_p.js]: Abstraction was kind of the point :) but nonetheless, I've removed it and updated unit tests and documentation

**Comment by jj22ee** [comment]
>>Hi, can you address @carolabadeer's comment: https://github.com/aws/aws-xray-sdk-node/pull/590#discussion_r1194387982

>Done and committed.

You may have addressed the comment below the comment I linked. `chai-as-promised` can be removed from the `package.json`/`package-lock.json`.

>packages/core/tsconfig.debug.json

Was this file intended to be removed when you moved your contributions to `sdk_contrib` folder, which I see you've added the same `tsconfig.debug.json` there? If so, can you remove this file.

**Comment by jasonterando** [comment]
> > > Hi, can you address @carolabadeer's comment: [#590 (comment)](https://github.com/aws/aws-xray-sdk-node/pull/590#discussion_r1194387982)
> 
> > Done and committed.
> 
> You may have addressed the comment below the comment I linked. `chai-as-promised` can be removed from the `package.json`/`package-lock.json`.
> 
> > packages/core/tsconfig.debug.json
> 
> Was this file intended to be removed when you moved your contributions to `sdk_contrib` folder, which I see you've added the same `tsconfig.debug.json` there? If so, can you remove this file.

Committed removal of chai-as-promised and tsconfig.debug.json

**Comment by jj22ee** [comment]
Can you look into the failing tests? There seems to be an issue with `activeFetch`.
Also, `packages/core/tsconfig.debug.json` wasn't actually removed in your latest commit.

**Comment by am29d** [comment]
Hey @jasonterando , appreciate the work you have put into this PR! Is there a chance I can help you move it forward, fixing tests or other work? It'd be great to push this PR to release. Let me know how I can assist. 

Best
Alex

**Comment by jasonterando** [comment]
> Hey @jasonterando , appreciate the work you have put into this PR! Is there a chance I can help you move it forward, fixing tests or other work? It'd be great to push this PR to release. Let me know how I can assist.
> 
> Best Alex

Hi Alex.  Yeah, I'm stuck on getting the test to run, trying to get lerna set up and running in Node 14

> Hey @jasonterando , appreciate the work you have put into this PR! Is there a chance I can help you move it forward, fixing tests or other work? It'd be great to push this PR to release. Let me know how I can assist.
> 
> Best Alex

Hi Alex,

You're the first person that's shown any appreciation for any of this.  I appreciate that.  Yeah, I'm stuck on the testing part of it.  Happy to work with somebody on getting this out, but it seems like it's one papercut after another, ex. the demand to "use const instead of var" even though var is used everywhere else in the code base.  I had it all working at one point and then  jj22ee insisted that I removed captureFetch, which broke all the testing in Node 14.  I messed around with trying to get Node 14 running locally in a container, getting lerna going, etc. and ran into version errors, etc. doing an npm install.  I can't even run mocha in the Docker Node 14 container, which is insisting I rename chai.js to .cjs.  

Just kind of gave up...

**Comment by jj22ee** [comment]
Hey @jasonterando, I don't mind the reintroduction of the generic `captureFetch` function. In this case, can you update its previous comment about the condition in which the global instance will be patched rather than the module. e.g. "in Node18+ the built-in Fetch instance will be patched rather than the node-fetch module".

**Comment by jasonterando** [comment]
I was doing another look at the code and remembered that I had to do this to get it working:
```js
const contextUtils = require('aws-xray-sdk-core/lib/context_utils');
const utils = require('aws-xray-sdk-core/lib/utils');
const logger = require('aws-xray-sdk-core/lib/logger');
```

I'm not sure if the links into aws-xray-sdk-core/lib are acceptable or not.  I'm thinking not.

The reason I had to go that route is that I was replicating what the http_p.js hook does, which is this at line 99:

```js
const parent = contextUtils.resolveSegment(contextUtils.resolveManualSegmentParams(options));
```
The method `resolveSegment` is availalbe via the aws-xray-sdk-core's exports, but `resolveManualSegmentParams` is not.  

Seems like I have three choices, if I want to mirror what's being done in the http patcher:

1. Add `resolveManualSegmentParams` to aws-xray-sdk-core's exports in aws-xray.js
2. Move the fetch support back to /lib/patchers
3. Leave the links to /lib and leave the fetch support in sdk_contrib

Thoughts?  (and thanks)

File: sdk_contrib/fetch/package.json
Comment (jj22ee) [inline sdk_contrib/fetch/package.json:3]: Update to 3.5.3

File: sdk_contrib/fetch/lib/fetch_p.js
Comment (jasonterando) [reply sdk_contrib/fetch/lib/fetch_p.js]: Done

File: sdk_contrib/fetch/package.json
Comment (jasonterando) [reply sdk_contrib/fetch/package.json]: Done

File: sdk_contrib/fetch/package.json
Comment (jj22ee) [inline sdk_contrib/fetch/package.json:17]: Can you update this to match the other script tests
`"test": "mocha --recursive --exit ./test/ -R spec && tsd",`

**Comment by jj22ee** [comment]
Thanks, just a few more small changes in addition to the new comments.

[A previous request](https://github.com/aws/aws-xray-sdk-node/pull/590#issuecomment-1861989339) was to remove `chai-as-promised` from `package.json/package-lock.json`. Looks like you instead removed it from `sdk_contrib/fetch/package.json` instead. Can you move this dependency back into `sdk_contrib`?

Also please remove the following files 
- `packages/core/tsconfig.debug.json`
- `.dockerignore`
- `sdk_contrib/fetch/package-lock.json` (usually we just keep the top level package-lock.json)
- `packages/core/package-lock.json`

File: sdk_contrib/fetch/package.json
Comment (jasonterando) [reply sdk_contrib/fetch/package.json:17]: Done

Comment (jasonterando) [reply sdk_contrib/fetch/package.json]: Done

File: sdk_contrib/fetch/lib/fetch_p.js
Comment (jj22ee) [inline sdk_contrib/fetch/lib/fetch_p.js:19]: Rename to `captureFetchGlobal`

File: sdk_contrib/fetch/test-d/index.test-d.ts
Comment (jj22ee) [inline sdk_contrib/fetch/test-d/index.test-d.ts:2]: `captureFetch` should be removed in this file.

**Comment by jasonterando** [comment]
Ugh... I'll have to look at the tests failing after the updated npm test script later tonight

**Comment by jasonterando** [comment]
I believe this line is failing:
`TypeError: Cannot stub non-existent property resolveManualSegmentParams`
because the updated aws-xray.js which adds the export for resolveManualSegmentParams is probably not being used in the pipeline.  Do we need to release that first before it's available for use by sdk_contrib, or is there some "trick" I can do in package.json to make this work?

**Comment by jasonterando** [comment]
I've gone ahead and put in a PR to add the resolveManualSegementParams export.  If this is the way we need to go on this, I can wait until this is "live" before picking the fetch merge back up.




    On Wednesday, January 24, 2024 at 03:06:35 PM CST, Jonathan Lee ***@***.***> wrote:  
 
 



Do we need to release that first before it's available for use by sdk_contrib, or is there some "trick" I can do in package.json to make this work?


Yeah, releasing "export for resolveManualSegmentParams" would be the more proper way of addressing this issue. The riskier and less ideal workaround would be to merge+release your PR, then hope that the Continuous Build Github Workflows will pass after the new release. We'll see if there's a better alternative.

**Comment by jj22ee** [comment]
@jasonterando 
New X-Ray SDK version is released, and I added a couple code changes to unblock your PR's unit tests
https://github.com/aws/aws-xray-sdk-node/blob/master/CHANGELOG.md#354

I looked into the segment generated by this PR, and I noticed that the Fetch Response information was missing in the segment.
```
"subsegments": [
    {
        "id": "22a15354bc8781eb",
        "name": "random-endpoint.com",
        "start_time": 1708993528.985,
        "end_time": 1708993529.44,
        "http": {
            "request": {
                "url": "",
                "method": "GET"
            },
            "response": {}
        },
        "namespace": "remote"
    }
]
```

I saw that this is because the `subsegment.addRemoteRequestData()` method was designed specifically for the built-in HTTP module Request/Response, and Fetch has a different Request/Response object. Therefore the data (request url/method and response status/content-length) must be obtained differently.

See how it is done for HTTP:
- https://github.com/aws/aws-xray-sdk-node/blob/4644c81171bfce63671ac5201cafe5a66f34f7d0/packages/core/lib/segments/attributes/subsegment.js#L252-L258
- https://github.com/aws/aws-xray-sdk-node/blob/4644c81171bfce63671ac5201cafe5a66f34f7d0/packages/core/lib/segments/attributes/remote_request_data.js#L16-L29

Do you have bandwidth anytime to implement extracting Fetch request/response data into subsegment and corresponding unit tests? Otherwise I could take a stab at it later.

**Comment by jasonterando** [comment]
@jj22ee  - yeah I can take a look at it, probably not until the weekend though, day job is hectic.  Should I work from fetch-pr?

**Comment by jj22ee** [comment]
You can still work from this PR's `fetch` branch. As for adding your own custom `addRemoteRequestData()` method for Fetch, you can add it to `fetch_p.js` to modify the subsegment's `http` property

**Comment by jasonterando** [comment]
Hi, just committed updates that write to subsegment.http.  One thing that wasn't clear was how traced is supposed to work.  It looks like RemoteRequestData.init sets request.traced to true if downstreamXRayEnabled) is true, but then traced gets subsequently moved to the subsegment in addRemoteRequestData.  In my code, I'm just setting subsegment.traced to true if downstreamXRayEnabled is true.  If that's wrong, and I need to do the "two-step", let me know.

File: sdk_contrib/fetch/lib/subsegment_fetch.js
Comment (jj22ee) [inline sdk_contrib/fetch/lib/subsegment_fetch.js:14]:
```suggestion
      url: (request.url) ? request.url.toString() : '',
```

**Comment by jj22ee** [comment]
Thanks for adding `subsegment_fetch.js`! Added one suggestion above.
Also, can you add the corresponding `subsegment_fetch.d.ts` file. After that, everything looks good to merge.

File: sdk_contrib/fetch/lib/subsegment_fetch.js
Comment (jasonterando) [reply sdk_contrib/fetch/lib/subsegment_fetch.js]: Ok, done.  I also put in a non-set check for method.  .d.ts file is added for Subsegment.addFetchRequestData

**Review by jj22ee — APPROVED** [review]
LGTM!

**Comment by jj22ee** [comment]
TYSM for the contribution @jasonterando!

**Comment by am29d** [comment]
Thanks a lot @jasonterando for the effort and @jj22ee for the review. Truly appreciate the merge ❤️

**Pull request merged**
